test_cmd/run_cmd_tests.sh

```diff
@@ -0,0 +1,210 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+DIR="$(dirname $0)"
+
+# Global counters
+EXECUTED=0
+FAILED=0
+
+
+JSONNET_BIN="${JSONNET_BIN:-../jsonnet}"
+
+# Display additional information such as successful tests (useful if they are slow).
+VERBOSE="${VERBOSE:-false}"
+
+separator() {
+    echo -e "\n-----------------------------\n"
+}
+
+
+check_file() {
+    TEST="$1"
+    RESULT_FILENAME="$2"
+    GOLDEN_FILENAME="$3"
+    if [ "$(< $RESULT_FILENAME)" != "$(< $GOLDEN_FILENAME)" ] ; then
+        FAILED=$((FAILED + 1))
+        if [ -z "$SUMMARY_ONLY" ]; then
+            echo -e "\e[31;1mFAIL\e[0m \e[1m(stdout mismatch)\e[0m: \e[36m$TEST\e[0m"
+            echo "Result output:"
+            cat "$RESULT_FILENAME"
+            echo "Expected output:"
+            cat "$GOLDEN_FILENAME"
+            echo "Diff:"
+            # Using git diff for pretty format and pretty colors
+            git --no-pager diff "$RESULT_FILENAME" "$GOLDEN_FILENAME"
+            # The output is often quite long, let's repeat the filename once more
+            # to avoid wasting time on looking for it
+            echo -e "\nTEST ABOVE: \e[36m$TEST\033[0m"
+            separator
+        fi
+        return 1
+    fi
+}
+
+do_test() {
+    TEST="$1"
+    EXPECTED_EXIT_CODE="$2"
+    shift 2
+
+    STDIN="${TEST}.stdin"
+    GOLDEN_STDOUT="${TEST}.golden.stdout"
+    GOLDEN_STDERR="${TEST}.golden.stderr"
+
+    OUT_DIR="out/${TEST}" 
+    STDOUT="${OUT_DIR}/stdout"
+    STDERR="${OUT_DIR}/stderr"
+
+    if [ ! -r "${STDIN}" ] ; then
+        STDIN="/dev/null"
+    fi
+
+    if [ ! -r "${GOLDEN_STDOUT}" ] ; then
+        GOLDEN_STDOUT="/dev/null"
+    fi
+
+    if [ ! -r "${GOLDEN_STDERR}" ] ; then
+        GOLDEN_STDERR="/dev/null"
+    fi
+
+
+    mkdir -p "${OUT_DIR}"
+    "${JSONNET_BIN}" "$@" >"${STDOUT}" 2>"${OUT_DIR}/stderr" <"${STDIN}"
+    TEST_EXIT_CODE="$?"
+    EXECUTED=$((EXECUTED + 1))
+
+    if [ "$TEST_EXIT_CODE" -ne "$EXPECTED_EXIT_CODE" ] ; then
+        FAILED=$((FAILED + 1))
+        if [ -z "$SUMMARY_ONLY" ]; then
+            echo -e "\e[31;1mFAIL\e[0m \e[1m(exit code)\e[0m: \e[36m$TEST\e[0m"
+            echo "This run's stdout:"
+            cat "$STDOUT"
+            echo "This run's stderr:"
+            cat "$STDERR"
+            echo "Actual exit code $TEST_EXIT_CODE, expected $EXPECTED_EXIT_CODE"
+            separator
+        fi
+        return 1
+    fi
+
+    if ! check_file "${TEST}" "${STDOUT}" "${GOLDEN_STDOUT}"; then
+        return 1
+    fi
+    
+    if ! check_file "${TEST}" "${STDERR}" "${GOLDEN_STDERR}"; then
+        return 1
+    fi
+}
+
+pushd "${DIR}"
+
+rm -rf out
+
+# And now the tests:
+
+# do_test <name> <exit_code> <args...>
+
+# Only check_file additional files if the test otherwise succeded, to avoid double-counting
+# failures.  All tests must write into out/$NAME/ and nowhere else.
+
+do_test "no_args" 1
```

Comment:
Can we extract the `do_test` part to a separate file? It will be easier to inspect them and add new ones without all of the machinery around them.